compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirMemberPropertyChecker.kt

```diff
@@ -5,32 +5,132 @@
+        val constructorToData = mutableMapOf<FirConstructor, PropertyInitializationInfo>().withDefault { PropertyInitializationInfo.EMPTY }
+
+        fun collectInfoFromGraph(
+            graph: ControlFlowGraph,
+            map: MutableMap<FirProperty, EventOccurrencesRange>,
+            acc: (EventOccurrencesRange, EventOccurrencesRange) -> EventOccurrencesRange,
+            delegatedConstructor: FirConstructor? = null,
+        ) {
+            val delegatedInfo =
+                if (delegatedConstructor != null && constructorToData.containsKey(delegatedConstructor)) {
+                    constructorToData.getValue(delegatedConstructor)
+                } else PropertyInitializationInfo.EMPTY
+
+            val data = PropertyInitializationInfoCollector(memberProperties.map { it.symbol }.toSet()).getData(graph)
+            val infoAtExitNode = data[graph.exitNode]?.get(NormalPath) ?: PropertyInitializationInfo.EMPTY
+
+            val info = when {
+                delegatedInfo.isEmpty() -> infoAtExitNode
+                infoAtExitNode.isEmpty() -> delegatedInfo
+                else -> {
+                    // NB: it's not merge, which is conducted at merging points, such as loop condition or when conditions.
+                    // Rather, delegated constructor call is the predecessor of the current constructor call, so we should accumulate.
+                    delegatedInfo.plus(infoAtExitNode)
+                }
+            }
+
+            if (graph.declaration is FirConstructor) {
+                constructorToData.putIfAbsent(graph.declaration as FirConstructor, info)
             }
+
+            for (property in memberProperties) {
+                if (map.containsKey(property)) {
+                    // Accumulation: && for class constructors, || for class's anonymous initializers and property initializations
+                    map[property] = acc.invoke(map[property]!!, info[property.symbol] ?: EventOccurrencesRange.ZERO)
+                } else {
+                    // Initial assignment.
+                    map[property] = info[property.symbol] ?: EventOccurrencesRange.ZERO
+                }
+            }
+        }
+
+        val constructorGraphs = declaration.constructors.mapNotNull { it.controlFlowGraphReference?.controlFlowGraph }
+        for (graph in constructorGraphs) {
+            var delegatedConstructor: FirConstructor? = null
+            val delegatedConstructorCall = (graph.declaration as? FirConstructor)?.delegatedConstructor
+            if (delegatedConstructorCall != null && delegatedConstructorCall.isThis) {
+                delegatedConstructor =
+                    (delegatedConstructorCall.calleeReference as? FirResolvedNamedReference)?.resolvedSymbol?.fir as? FirConstructor
+            }
+            collectInfoFromGraph(graph, initializedInConstructor, EventOccurrencesRange::or, delegatedConstructor)
```

Comment:
Trying to understand what the use of `::or` here means. I thought you are trying to do `&&` for constructors. Or the name of this operation is unintuitive?

Reply:
Oh, I forgot to update comments about accumulation operation after changing the theme of map from boolean initialization to event occurrences.

I have to admit the name of `::or` is somewhat unintuitive, but that one is used everywhere so I'd not change that. You're right that we want to do `&&` for constructors: a property should be initialized at all constructors, but from event occurrence viewpoint, that's `join`ining of all occurrences, which is named `or`. :\

Reply:
Given two unrelated constructors A and B, the intuition I use to understand this is that the property initialization could happen in constructor A `or` constructor B, but not both.

Reply:
The correct interpretation is, after `or`-ing or event occurrences in those constructors, which computes the lower and upper bound of event occurrences, getting `EXACTLY_ONCE` (or [1, 1]) means the property is initialized at all constructors. Otherwise, there is a hole where the property can not be initialized.